langchain4j-core/src/main/java/dev/langchain4j/agent/tool/ToolSpecifications.java

```diff
@@ -98,7 +95,72 @@ static Iterable<JsonSchemaProperty> toJsonSchemaProperties(Parameter parameter)
             return removeNulls(STRING, enums((Object[]) type.getEnumConstants()), description);
         }
 
-        return removeNulls(OBJECT, description); // TODO provide internals
+        return removeNulls(OBJECT, schema(type), description);
+    }
+
+    private static JsonSchemaProperty schema(Class<?> structured){
+        //TODO: Externalize the depth
+        return schema(structured, new HashSet<>(), 3);
```

Comment:
Why depth is 3?

Reply:
This is to keep a check on how deep we recursively generate schema. 
Wanted this be configurable, but not sure how to do in this framework

Reply:
In theory, the object graph can be deeper than 3 levels. Recursion is stopped by using the map of visited classes. Do we really need this depth checks?